### ISSUE TITLE
Komga: Faster downloads

### ISSUE THREAD

**Issue by TheHardew** [tachiyomiorg/tachiyomi-extensions#12025]
### Describe your suggested feature

Currently, when downloading Komga Manga I get around 1-3 MiB/s.
I thought Tachiyomi might be rate limiting the connection, but I couldn't find any such settings.
I believe what happens is Tachiyomi downloads manga one page at a time.
When using REST API with curl I get simillar speeds.
A way to make the downloads faster would be to: 

 - Download the whole chapter instead of one page at a time. Especially useful if Tachiyomi is set to archive the pages it downloads. This way we can avoid Komga uncompressing the pages just for them to be compressed again.
 - Download multiple pages in parallel. With curl it allowed me to reach speeds comparable to the previous method.

The [FAQ says](https://tachiyomi.org/help/faq/#downloads) that parallel downloads are not supporeted for a single source in order to prevent ip bans, but if one hosts their own server, that's hardly an issue. Maybe there should be a toggle in the extension settings.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] If this is an issue with the app itself, I should be opening an issue in the [app repository](https://github.com/tachiyomiorg/tachiyomi/issues/new/choose).
- [X] I have updated the app to version **[0.13.4](https://github.com/tachiyomiorg/tachiyomi/releases/latest)**.
- [X] I will fill out all of the requested information in this form.

**Comment by arkon** [comment]
> I believe what happens is Tachiyomi downloads manga one page at a time.

It downloads 5 pages at a time.

> When using REST API with curl I get simillar speeds. [...] With curl it allowed me to reach speeds comparable to the previous method.

So is curl faster or not? These two statements seem to conflict.

> Download the whole chapter instead of one page at a time. Especially useful if Tachiyomi is set to archive the pages it downloads. This way we can avoid Komga uncompressing the pages just for them to be compressed again.

The app doesn't support downloading arbitrary archives as chapters since they need to follow the apps' expected format.

**Comment by TheHardew** [comment]
I modified Tachiyomi source to download multiple chapters at once from the same source  (by using `flatMap` instead of `concatMap`) and it did establish more connections and the interface speed was higher. Altough frankly I don't quite understand how it works, since I can't find in the documentation anything about `List.flatMap` taking 2 arguments and I don't know kotlin. In any case, there is a way to make it faster.

**Comment by AntsyLich** [comment]
> In any case, there is a way to make it faster.

By destroying the source

**Comment by Seishirou101** [comment]
What endpoint did you use to download? You can look at the ext code to see what the app uses

**Comment by TheHardew** [comment]
I'm sorry, I'm not sure what you mean. Komga extension 1.2.39, Tachiyomi 0.13.4. 

I opened the issue in the extensions, but looking and this now the downloading happens in the main app so it might not be the right place.

**Comment by Seishirou101** [comment]
The api endpoint you're using for komga to compare to the app

**Comment by TheHardew** [comment]
https://demo.komga.org/swagger-ui/index.html?configUrl=/v3/api-docs/swagger-config

This is the api I used. Of course I tested it on my own server. But even when compiling the app to download chapters in parallel I saw a speed up

**Comment by TheHardew** [comment]
`curl -X GET 'pi.hole/komga/api/v1/books/08VP0PGT2KSND/pages/1' -u 
EMAIL:PASSWORD -sO` (run this with a bash shell)  

`seq 400 | parallel curl -X GET 'pi.hole/komga/api/v1/books/08VP0PGT2KSND/pages/{}' -u 
EMAIL:PASSWORD -sO`

**Comment by TheHardew** [comment]
You might need to remove `/komga` from this btw, it's just there for my reverse proxy, but it requires the server to be started with ` --server.servlet.context-path="/komga"`

**Comment by Seishirou101** [comment]
Okay, I wanted to make sure you were using the same endpoint. are you testing this on the same device as tachiyomi? I'm getting roughly the same speed with curl, as arkon said, it downloads 5 pages at a time.

**Comment by TheHardew** [comment]
Yes, I switched to testing on the phone.

**Comment by Seishirou101** [comment]
I'm not sure why the discrepancy of download speeds then, you can come on discord and share your results to see if someone can find the discrepancy just in case its a bug, but from my testing they are similar. As there is no issue with the extension I will close this. You can open an issue on the main repo to request if self hosted extensions can bypass the parallel download rule.
